#Weeks 8 & 9 - Pandas
In this homework assignment, you will explore and analyze a public dataset of your choosing. Since this assignment is “open-ended” in nature, you are free to expand upon the requirements below. However, you must meet the minimum requirments as indicated in each section.

You must use Pandas as the primary tool to process your data.

The preferred method for this analysis is in a .ipynb file. Feel free to use whichever platform of your choosing.

https://www.youtube.com/watch?v=inN8seMm7UI (Getting started with Colab).

Your data should need some "work", or be considered "dirty". You must show your skills in data cleaning/wrangling.

Some data examples:
• https://www.data.gov/

• https://opendata.cityofnewyork.us/

• https://datasetsearch.research.google.com/

• https://archive.ics.uci.edu/ml/index.php

Resources:
• https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html

• https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html

# Headings or comments
You are required to make use of comments, or headings for each section. You must explain what your code is doing, and the results of running your code. Act as if you were giving this assignment to your manager - you must include clear and descriptive information for each section.

# Introduction
In this section, please describe the dataset you are using. Include a link to the source of this data. You should also provide some explanation on why you choose this dataset.

#Data Exploration

The chosen data set is from the Integrated Global Radiosonde Archive (IGRA) on the National Oceanic and Atmospheric Administration's website (NOAA).  The available file types are .txt files.  These file contain data obtained from radiosondes, which are devices used to record pressure, temperature, relative humidity, wind speed, and wind direction.  These radiosondes are attached to weather balloons and record the data at varying heights within Earth's atmosphere. The data set that I chose was for Puerto Rico during summmer months.  Prior to importing my data set to Python, I imported my data set to Microsoft Excel to narrow down the large data set as there were over 6 million recordings dating back to 1946. I narrowed down the data set to only 1800 recordings during June 2021.

In [ ]:
import pandas as pd

# Data Exploration
#Import your dataset into your .ipynb, create dataframes, and explore your data.
#Include: Summary statistics means, medians, quartiles, Missing value information
#Any other relevant information about the dataset.
df = pd.read_csv('PR.txt')

df

#RQM00078526 2021 06 05 00 2307  248 ncdc-nws ncdc-nws  184317  -659919
0     21     0 101592B    3   271B  770    44   123 ...                     
1     20    26 100145   130B  255B  822    32   112 ...                     
2     10    29 100000   143B  253B  825    32   111 ...                     
3     20   142  96289   475B  229B  825    31   103 ...                     
4     10   306  92500   825B  205B  832    29    94 ...                     
...                                                 ...                     
1794  30 10900  -9999 33329 -9999 -9999 -9999   106 ...                     
1795  30 11000  -9999 33699 -9999 -9999 -9999    95 ...                     
1796  30 11100  -9999 34083 -9999 -9999 -9999    93 ...                     
1797  30 11200  -9999 34462 -9999 -9999 -9999    61 ...                     
1798  30 11300  -9999 34817 -9999 -9999 -9999    42 ...                     

[1799 rows x 1 columns]

#Narrowing Down
Despite narrowing down my data, I still had so much more data that I needed. My plan was to narrow down my data to one given day, namely June 8th, 2021, to compute the value of Convective Available Potential Energy (CAPE) based on pressure, temperature, and relative humidity values.  I used the drop() method to delete unnecessary rows by index, including rows corresponding to other days and rows that had invalid entries.  Luckily, the invalid entry rows were grouped together with '-9999' as an entry and the data is already sorted in decreasing pressure values.

In [ ]:
#drop unncessary rows
df = df.drop(df.index[1453:1799])
df = df.drop(df.index[0:1340])

df

#RQM00078526 2021 06 05 00 2307  248 ncdc-nws ncdc-nws  184317  -659919
1340  21     0 101586B    3   270B  790    40    92 ...                     
1341  10    30 100000   142B  258B  812    35    82 ...                     
1342  20   140  96228   480B  227B  917    14    85 ...                     
1343  20   217  94585   631B  218B  865    24    85 ...                     
1344  10   302  92500   825B  205B  898    17    83 ...                     
...                                                 ...                     
1448  20 10118    783 32824B -360B   10   382    96 ...                     
1449  20 10257    718 33427B -376B   10   377    86 ...                     
1450  10 10325    700 33598B -349B   10   385    86 ...                     
1451  20 10335    694 33655B -336B   10   389    86 ...                     
1452  20 10543    621 34437B -309A   10   398    99 ...                     

[113 rows x 1 columns]

#Renaming and Creating Columns
From the remaining rows, the header row data was also not needed, but needed to be renamed.  However, the header row was also just one column.  I renamed the column as 'All Columns' to eventually create new columns using the original data.

In [ ]:
#renaming header column
df = df.rename(columns={"#RQM00078526 2021 06 05 00 2307  248 ncdc-nws ncdc-nws  184317  -659919" : "All Columns"})

df

All Columns
1340  21     0 101586B    3   270B  790    40    92 ...
1341  10    30 100000   142B  258B  812    35    82 ...
1342  20   140  96228   480B  227B  917    14    85 ...
1343  20   217  94585   631B  218B  865    24    85 ...
1344  10   302  92500   825B  205B  898    17    83 ...
...                                                 ...
1448  20 10118    783 32824B -360B   10   382    96 ...
1449  20 10257    718 33427B -376B   10   377    86 ...
1450  10 10325    700 33598B -349B   10   385    86 ...
1451  20 10335    694 33655B -336B   10   389    86 ...
1452  20 10543    621 34437B -309A   10   398    99 ...

[113 rows x 1 columns]

#Splitting the Columns
A big difficulty of mine was trying to split the data using a blank space as the delimiter. However, because of the varying digits in the data, some rows had more spaces between data values and did not split properly. The way that the data values were recorded was very confusing also and included string-letter characters.  Instead, I used the slice() method to split my data and create columns with fixed width. I then removed the original column from the data frame.

In [ ]:
#create new columns from the original column by slicing
df['LVLTYPE'] = df['All Columns'].str.slice(stop=2)
df['ETIME'] = df['All Columns'].str.slice(start=4, stop = 8)
df['PRESS'] = df['All Columns'].str.slice(start=9, stop = 15)
df['GPH'] = df['All Columns'].str.slice(start=16, stop = 21)
df['TEMP'] = df['All Columns'].str.slice(start=23, stop=27)
df['RH'] = df['All Columns'].str.slice(start=28, stop = 33)
df['DPDP'] = df['All Columns'].str.slice(start=34, stop = 39)
df['WDIR'] = df['All Columns'].str.slice(start=40, stop=45)
df['WSPD'] = df['All Columns'].str.slice(start=46, stop = 51)

df

All Columns LVLTYPE ETIME   PRESS  \
1340  21     0 101586B    3   270B  790    40    92 ...      21     0  101586   
1341  10    30 100000   142B  258B  812    35    82 ...      10    30  100000   
1342  20   140  96228   480B  227B  917    14    85 ...      20   140   96228   
1343  20   217  94585   631B  218B  865    24    85 ...      20   217   94585   
1344  10   302  92500   825B  205B  898    17    83 ...      10   302   92500   
...                                                 ...     ...   ...     ...   
1448  20 10118    783 32824B -360B   10   382    96 ...      20  0118     783   
1449  20 10257    718 33427B -376B   10   377    86 ...      20  0257     718   
1450  10 10325    700 33598B -349B   10   385    86 ...      10  0325     700   
1451  20 10335    694 33655B -336B   10   389    86 ...      20  0335     694   
1452  20 10543    621 34437B -309A   10   398    99 ...      20  0543     621   

        GPH  TEMP     RH   DPDP   WDIR   WSPD  
1340      3   270    790     40     92     41  
1341    142   258    812     35     82     91  
1342    480   227    917     14     85    100  
1343    631   218    865     24     85    114  
1344    825   205    898     17     83    109  
...     ...   ...    ...    ...    ...    ...  
1448  32824  -360     10    382     96    253  
1449  33427  -376     10    377     86    276  
1450  33598  -349     10    385     86    242  
1451  33655  -336     10    389     86    227  
1452  34437  -309     10    398     99    237  

[113 rows x 10 columns]

The original column still existed in the data frame and still had all of the original data, so I dropped it.

In [ ]:
#drop 'All Columns'
df = df.drop(columns = ['All Columns'])

df

LVLTYPE ETIME   PRESS    GPH  TEMP     RH   DPDP   WDIR   WSPD
1340      21     0  101586      3   270    790     40     92     41
1341      10    30  100000    142   258    812     35     82     91
1342      20   140   96228    480   227    917     14     85    100
1343      20   217   94585    631   218    865     24     85    114
1344      10   302   92500    825   205    898     17     83    109
...      ...   ...     ...    ...   ...    ...    ...    ...    ...
1448      20  0118     783  32824  -360     10    382     96    253
1449      20  0257     718  33427  -376     10    377     86    276
1450      10  0325     700  33598  -349     10    385     86    242
1451      20  0335     694  33655  -336     10    389     86    227
1452      20  0543     621  34437  -309     10    398     99    237

[113 rows x 9 columns]

In [ ]:
#change data frame to floats
df = df.astype(float)

df

LVLTYPE  ETIME     PRESS      GPH   TEMP     RH   DPDP  WDIR   WSPD
1340     21.0    0.0  101586.0      3.0  270.0  790.0   40.0  92.0   41.0
1341     10.0   30.0  100000.0    142.0  258.0  812.0   35.0  82.0   91.0
1342     20.0  140.0   96228.0    480.0  227.0  917.0   14.0  85.0  100.0
1343     20.0  217.0   94585.0    631.0  218.0  865.0   24.0  85.0  114.0
1344     10.0  302.0   92500.0    825.0  205.0  898.0   17.0  83.0  109.0
...       ...    ...       ...      ...    ...    ...    ...   ...    ...
1448     20.0  118.0     783.0  32824.0 -360.0   10.0  382.0  96.0  253.0
1449     20.0  257.0     718.0  33427.0 -376.0   10.0  377.0  86.0  276.0
1450     10.0  325.0     700.0  33598.0 -349.0   10.0  385.0  86.0  242.0
1451     20.0  335.0     694.0  33655.0 -336.0   10.0  389.0  86.0  227.0
1452     20.0  543.0     621.0  34437.0 -309.0   10.0  398.0  99.0  237.0

[113 rows x 9 columns]

#Computing Convective Available Potential Energy (CAPE)
For my intents and purposes, the only data that was necessary was pressure (in hectopascals), temperature (in degrees celsius) and the relative humidity (as a floating point value between 0 and 1). As these values were not in the correct units, I converted these columns, reassigned them with the correct values, then dropped unneeded columns.


In [ ]:
#The pressure needed to be coverted by dividing by 100
df['PRESS'] = df['PRESS'] / 100

#The temperature needed to be converted by dividing by 10
df['TEMP'] = df['TEMP'] / 10

#The temperature needed to be converted by dividing by 1000
df['RH'] = df['RH'] /1000

#Dropped the unneeded columns
df = df.drop(columns = ['LVLTYPE', 'ETIME', 'GPH', 'DPDP', 'WDIR', 'WSPD'])

df

PRESS  TEMP        RH
1340  10.1586  2.70  0.000790
1341  10.0000  2.58  0.000812
1342   9.6228  2.27  0.000917
1343   9.4585  2.18  0.000865
1344   9.2500  2.05  0.000898
...       ...   ...       ...
1448   0.0783 -3.60  0.000010
1449   0.0718 -3.76  0.000010
1450   0.0700 -3.49  0.000010
1451   0.0694 -3.36  0.000010
1452   0.0621 -3.09  0.000010

[113 rows x 3 columns]

In [ ]:
print(df.describe())

            PRESS        TEMP          RH
count  113.000000  113.000000  113.000000
mean     3.624616   -2.498407    0.000343
std      2.944922    2.933505    0.000327
min      0.062100   -7.200000    0.000010
25%      0.700000   -5.050000    0.000015
50%      3.454000   -2.590000    0.000268
75%      5.750600   -0.090000    0.000639
max     10.158600    2.700000    0.000999


Sort your data based on multiple variables.

Filter your data based on some condition.

Convert all the string values to upper or lower cases in one column.

Check whether numeric values are present in a given column of your dataframe.

Group your dataset by one column, and get the mean, min, and max values by group.

Groupby()
agg() or .apply()
Group your dataset by two columns and then sort the aggregated results within the groups.
You are free (and should) to add on to these questions.

In [ ]:
!pip install MetPy


from metpy.calc import cape_cin, dewpoint_from_relative_humidity, parcel_profile
from metpy.units import units
# pressure
p = list(df['PRESS']) * units.hPa

# temperature
T = list(df['TEMP']) * units.degC
# relative humidity
rh = list(df['RH']) * units.dimensionless
# calculate dewpoint
Td = dewpoint_from_relative_humidity(T, rh)
# compture parcel temperature
prof = parcel_profile(p, T[0], Td[0]).to('degC')
# calculate surface based CAPE/CIN
cape_cin(p, T, Td, prof)

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 400.5/400.5 kB 4.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 294.0/294.0 kB 7.6 MB/s eta 0:00:00


(2173.6401504440114 <Unit('joule / kilogram')>,
 -0.1266500439938402 <Unit('joule / kilogram')>)

# Conclusions
After exploring your dataset, provide a short summary of what you noticed from this dataset. What would you explore further with more time?